### PR TITLE
refactor(es/transforms/compat): visit_mut for class constructor

### PR DESCRIPTION
**Pull request by RiESAEX**
for #2370

### CODE REVIEW

File: crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs [hunk 1]
```diff
@@ -319,16 +318,16 @@ pub(super) enum SuperFoldingMode {
 }
 
 /// TODO: VisitMut
```
Comment (kwonoj) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs:320]: may remove comments?

File: crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs [hunk 2]
```diff
@@ -337,41 +336,42 @@ impl Fold for ConstructorFolder<'_> {
...
                     let this_super = self.cur_this_super.take().unwrap();
                     self.cur_this_super = old;
 
                     e.args.insert(0, this_super.as_arg());
-
-                    CallExpr {
-                        span: e.span,
-                        callee: e
-                            .callee
+                    e.callee.map_with_mut(|callee| {
```
Comment (kwonoj) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs:345]: this is q for @kdy1 : when do we recommend direct assignment vs. map_with_mut? I used direct assignment in most cases mostly not knowing map_with_mut.

File: crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs [hunk 3]
```diff
@@ -388,92 +388,99 @@ impl Fold for ConstructorFolder<'_> {
...
-        let arg = stmt.arg.fold_with(self);
-
-        let arg = Some(Box::new(make_possible_return_value(
-            ReturningMode::Returning {
-                mark: self.mark,
-                arg,
-            },
-        )));
+        stmt.arg.visit_mut_with(self);
+        // let arg = stmt.arg.to_owned();
```
Comment (kwonoj) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs:467]: unused comments?

File: crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs [hunk 4]
```diff
@@ -494,7 +501,7 @@ impl Fold for ConstructorFolder<'_> {
                                 ]
                             } else {
                                 let mut call_args = vec![ThisExpr { span: DUMMY_SP }.as_arg()];
-                                call_args.extend(args);
+                                call_args.extend(args.clone().into_iter());
```
Comment (kwonoj) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs:504]: are these need clone or if we can take?

File: crates/swc_ecma_transforms_compat/src/es2015/classes/mod.rs [hunk 5]
```diff
@@ -461,10 +463,10 @@ where
                 constructor.unwrap_or_else(|| default_constructor(super_class_ident.is_some()));
 
             // Rename variables to avoid conflicting with class name
-            constructor.body = constructor.body.fold_with(&mut VarRenamer {
+            constructor.body = constructor.body.fold_with(&mut as_folder(VarRenamer {
```
Comment (kwonoj) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/mod.rs:466]: curios if these still need fold, or can do `ctor.body.visit_mut` instead?

**Comment by RiESAEX** [comment]
@kwonoj Thanks!
I update most of them, excpet case of direct assignment vs. map_with_mut

File: crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs [hunk 6]
```diff
@@ -388,92 +387,98 @@ impl Fold for ConstructorFolder<'_> {
...
                 _ => {}
             },
             _ => {}
         }
 
-        let expr = match self.handle_super_assignment(expr) {
-            Ok(v) => v,
-            Err(v) => v.fold_children_with(self),
-        };
+        expr.map_with_mut(|expr| {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs:399]: `map_with_mut` is same as using `Fold`.
(It's `memmove` anyway.)

File: crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs [hunk 7]
```diff
@@ -494,7 +499,7 @@ impl Fold for ConstructorFolder<'_> {
                                 ]
                             } else {
                                 let mut call_args = vec![ThisExpr { span: DUMMY_SP }.as_arg()];
-                                call_args.extend(args);
+                                call_args.extend(args.take().into_iter());
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs:502]: `.into_iter()` seems redundant.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks! I found some tiny issues.

File: crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs [hunk 8]
```diff
@@ -251,7 +258,7 @@ impl ConstructorFolder<'_> {
                 }));
                 // _getPrototypeOf(Foo.prototype)
                 let get_proto = Box::new(Expr::Call(CallExpr {
-                    span,
+                    span: span.take(),
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2015/classes/constructor.rs:261]: Can you change this to `span: *span`?
`Span` implements `Copy` so `.take()` is redundant.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Almost LGTM, thanks!
I found one tiny issue about `Span`.

**Review by kdy1 — APPROVED** [review]
LGTM/ Thank you!

**Pull request merged**